### PR TITLE
[Fix #23] Show an error when creating an event with an end date after a start date

### LINKED ISSUE #23: Event can have an end date < start date
When creating an event, the end date can be set to be before the start date. This, of course, is nonsense. 😀 

screen shot 2016-10-10 at 23 07 42

### PR DESCRIPTION
**Pull request by alahab2**
This change checks that when creating an event `end_on` date cannot be before a `start_on` date.

---

**Before submitting, check that:**

 - [x] You have added (passing) tests for your code.
 - [x] You have written good* commit messages.
 - [x] You have squashed relevant commits together.
 - [x] You have ensured that RuboCop is passing.
 - [x] Your PR relates to one subject, with a clear title and
       description using complete, grammatically correct sentences.

---

*If you do not know what makes a good commit message, or why it
is important, please refer to these resources: [1][1], [2][2], [3][3].

[1]: https://robots.thoughtbot.com/5-useful-tips-for-a-better-commit-message
[2]: http://chris.beams.io/posts/git-commit/
[3]: https://github.com/blog/1943-how-to-write-the-perfect-pull-request

### CODE REVIEW

File: app/models/event.rb [hunk 1]
```diff
@@ -6,6 +6,13 @@ class Event < ActiveRecord::Base
   validates :start_on, presence: true
   validates :end_on, presence: true
   validates :name, presence: true
+  validate :validate_end_on_before_start_on
+
+  def validate_end_on_before_start_on
+    if end_on && start_on
```
Comment (Drenmi) [inline app/models/event.rb]: This check is not needed, since we're already validating presence of these two. 😀

File: app/models/event.rb [hunk 2]
```diff
@@ -6,6 +6,13 @@ class Event < ActiveRecord::Base
   validates :start_on, presence: true
   validates :end_on, presence: true
   validates :name, presence: true
+  validate :validate_end_on_before_start_on
+
+  def validate_end_on_before_start_on
```
Comment (Drenmi) [inline app/models/event.rb]: Let's name this something simpler, like maybe `#start_before_end`, then our DSL will read `validate :start_before_end` WDYT?

File: app/controllers/events_controller.rb [hunk 3]
```diff
@@ -15,7 +15,7 @@ def create
     if @event.save
       redirect_to events_path, notice: "Event was successfully created."
     else
-      redirect_to events_path, alert: "Failed to save event."
+      redirect_to events_path, alert: @event.errors.full_messages.to_sentence
```
Comment (Drenmi) [inline app/controllers/events_controller.rb]: This is awesome! We should do this to all our controller error paths.

File: spec/models/event_spec.rb [hunk 4]
```diff
@@ -13,4 +13,10 @@
 
     it { expect(described_class.search("event")).to contain_exactly(match_beginning, match_middle, match_end) }
   end
+
+  describe "#validate_end_on_before_start_on" do
+    it "does not allow end date to be before start date" do
+      expect(FactoryGirl.build(:event, name: "End Date before Start Date", start_on: 1.year.ago + 1.day, end_on: 1.year.ago)).to_not be_valid
```
Comment (Drenmi) [inline spec/models/event_spec.rb]: Let's move the building into a `let` block, and remove the description to `it` so we use the same style as our other tests. 😀

Comment (alahab2) [inline spec/models/event_spec.rb]: I tried `let` initially, but as the input is invalid the `let` is not being created and it throws an error.

File: app/models/event.rb (hunk 2 above)
Comment (alahab2) [inline app/models/event.rb]: 👍

File: spec/models/event_spec.rb (hunk 4 above)
Comment (Drenmi) [inline spec/models/event_spec.rb]: Hm. That is strange. The record shouldn't be validated until we try to save it, but .build should just build it. I'll give it a try tomorrow. 😀

File: app/models/event.rb (hunk 1 above)
Comment (alahab2) [inline app/models/event.rb]: Fixed everything except for this one. As soon as I remove `if end_on && start_on` I get `undefined method `<' for nil:NilClass`.

Comment (Drenmi) [inline app/models/event.rb]: Reading up a bit on this, there's no guaranteed order that validations will be run. I'll check how we handle this case in Tinkerbox. 😀

Comment (Drenmi) [inline app/models/event.rb]: After discussing this with my colleagues, we agree this is the best way. 😀 

We think it's to use the `#add` method on the errors object, though:

```
errors.add(:base, "Start date must be before end date")
```

**Comment by Drenmi** [comment]
🇰🇿

**Comment by alahab2** [comment]
Latest comments addressed. 😃

**Comment by Drenmi** [comment]
Squash commits and :shipit: ! 😄

**Pull request merged**